src/oumi/core/analyze/dataset_analyzer.py

```diff
@@ -100,30 +160,134 @@ def analyze_dataset(self) -> dict[str, Any]:
 
         logger.info(f"Analyzing {conversations_to_analyze} conversations")
 
-        # Step 1: Per-sample (message) level analysis
-        logger.info("Step 1: Computing per-sample (message) level analysis...")
+        # Step 1: Per-message level analysis
+        logger.info("Step 1: Computing message metrics...")
+
+        dataset_analysis_result = self._compute_message_metrics()
+
+        return dataset_analysis_result
+
+    def _compute_message_metrics(self) -> DatasetAnalysisResult:
+        """Compute metrics for all messages in the dataset.
+
+        Returns:
+            DatasetAnalysisResult: Structured results containing message-level analysis
+            with metadata about the analysis scope and individual message results.
+        """
+        total_conversations = len(self.dataset)
+
+        # Apply conversation limit if specified
+        max_conversations = self.config.sample_count
+
+        if max_conversations is not None:
+            if max_conversations <= 0:
+                raise ValueError(
+                    f"sample_count must be positive, got {max_conversations}. "
+                    "Use None to analyze all conversations."
+                )
+            conversations_to_analyze = min(total_conversations, max_conversations)
+            logger.info(
+                f"Limiting analysis to first {max_conversations} "
+                f"conversations (dataset has {total_conversations} total)"
+            )
+        else:
+            conversations_to_analyze = total_conversations
+
+        logger.info(
+            "Analyzing %d conversations for message-level metrics",
+            conversations_to_analyze,
+        )
+
+        # Collect all message analysis results
+        message_results = []
+
+        # Use tqdm for progress monitoring
+        with tqdm(
+            total=conversations_to_analyze,
+            desc=f"Analyzing {self.dataset_name}",
+            unit="conv",
+        ) as pbar:
+            for conv_idx in range(conversations_to_analyze):
+                conversation = self.dataset.conversation(conv_idx)
+
+                for msg_idx, message in enumerate(conversation.messages):
+                    message_result = self._compute_per_message_metrics(
+                        message, conv_idx, msg_idx, conversation
+                    )
+                    message_results.append(message_result)
+
+                pbar.update(1)
```

Comment:
This can be slightly cleaner. You shouldn't need to keep track of the progress bar if you have TQDM handle the iterator for you.

```suggestion
        for conv_idx in tqdm(range(conversations_to_analyze)
            desc=f"Analyzing {self.dataset_name}",
            unit="conv",
        ):
            conversation = self.dataset.conversation(conv_idx)
            for msg_idx, message in enumerate(conversation.messages):
                message_result = self._compute_per_message_metrics(
                    message, conv_idx, msg_idx, conversation
                )
                message_results.append(message_result)
```